packages/lodestar/src/network/encoders/response.ts

```diff
@@ -0,0 +1,107 @@
+import {IResponseChunk} from "./interface";
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {ILogger} from "@chainsafe/lodestar-utils/lib/logger";
+import {Method, MethodResponseType, Methods, ReqRespEncoding, RpcResponseStatus} from "../../constants";
+import {decode, encode} from "varint";
+import {encodeResponseStatus, getCompressor, getDecompressor} from "./utils";
+import BufferList from "bl";
+import {ResponseBody} from "@chainsafe/lodestar-types";
+
+export function eth2ResponseEncode(
+  config: IBeaconConfig, logger: ILogger, method: Method, encoding: ReqRespEncoding
+): (source: AsyncIterable<IResponseChunk>) => AsyncIterable<Buffer> {
+  return (source => {
+    return (async function*() {
+      const type = Methods[method].responseSSZType(config);
+      let compressor = getCompressor(encoding);
+      if(!type) {
+        return;
+      }
+      for await (const chunk of source) {
+        if(chunk.status !== 0) {
+          yield encodeResponseStatus(chunk.status);
+          break;
+        }
+        //yield status
+        yield encodeResponseStatus(chunk.status);
+        let serializedData: Buffer;
+        try {
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const serialized = type.serialize(chunk.body as any);
+          serializedData = Buffer.from(serialized.buffer, serialized.byteOffset, serialized.length);
+        } catch (e) {
+          logger.warn(`Failed to ssz serialize chunk of method ${method}. Error: ${e.message}`); 
+        }
+        //yield encoded ssz length
+        yield Buffer.from(encode(serializedData.length));
+        //yield compressed or uncompressed data chunks
+        yield* compressor(serializedData);
+        //reset compressor
+        compressor = getCompressor(encoding);
```

Comment:
do we have `reset` method like the decompressor and reuse the compressor?

Reply:
not really since compressor is just function, we can convert it to sam class later on, but we can do that once we replace snappy-stream with our implementation using snappyjs